spring-kafka/src/main/java/org/springframework/kafka/AsyncKafkaTemplate.java

```diff
@@ -0,0 +1,229 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.Iterator;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.internals.RecordHeader;
+
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.context.SmartLifecycle;
+import org.springframework.kafka.core.KafkaTemplate;
+import org.springframework.kafka.core.ProducerFactory;
+import org.springframework.kafka.listener.BatchMessageListener;
+import org.springframework.kafka.listener.GenericMessageListenerContainer;
+import org.springframework.kafka.support.KafkaHeaders;
+import org.springframework.kafka.support.SendResult;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.util.Assert;
+import org.springframework.util.concurrent.ListenableFuture;
+import org.springframework.util.concurrent.SettableListenableFuture;
+
+/**
+ * Async send/receive template.
+ *
+ * @param <K> the key type.
+ * @param <V> the outbound data type.
+ * @param <R> the reply data type.
+ *
+ * @author Gary Russell
+ * @since 2.1.3
+ *
+ */
+public class AsyncKafkaTemplate<K, V, R> extends KafkaTemplate<K, V> implements BatchMessageListener<K, R>,
+		InitializingBean, SmartLifecycle {
+
+	private static final long DEFAULT_REPLY_TIMEOUT = 5000L;
+
+	private final GenericMessageListenerContainer<K, R> replyContainer;
+
+	private final ConcurrentMap<String, RequestReplyFuture<K, V, R>> futures = new ConcurrentHashMap<>();
+
+	private TaskScheduler scheduler;
+
+	private boolean running;
+
+	private int phase;
+
+	private boolean autoStartup;
+
+	private long replyTimeout = DEFAULT_REPLY_TIMEOUT;
+
+	public AsyncKafkaTemplate(ProducerFactory<K, V> producerFactory,
+			GenericMessageListenerContainer<K, R> replyContainer) {
+		this(producerFactory, replyContainer, false);
+	}
+
+	public AsyncKafkaTemplate(ProducerFactory<K, V> producerFactory,
+			GenericMessageListenerContainer<K, R> replyContainer, boolean autoFlush) {
+		super(producerFactory, autoFlush);
+		this.replyContainer = replyContainer;
+		this.replyContainer.setupMessageListener(this);
+	}
+
+	protected void setTaskScheduler(TaskScheduler scheduler) {
+		this.scheduler = scheduler;
+	}
+
+	protected void setReplyTimeout(long replyTimeout) {
+		Assert.isTrue(replyTimeout >= 0, "'replyTimeout' must be >= 0");
+		this.replyTimeout = replyTimeout;
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		Assert.notNull(this.scheduler, "A scheduler is required for timeouts");
```

Comment:
Why just don't create an internal instance then?

It doesn't look correct to allow to create this `AsyncKafkaTemplate` without `TaskScheduler`, but then require it later.